Header error

little-dude opened the pull request:
Here is my attempt at fixing #67.

I also added a few clippy lints but I can move this to another PR if you prefer.

I noticed you use `try!` instead of `?`. I can either follow this convention or make changes in the crates to use `?` instead.

little-dude commented:
The build failure is due to travis:

```
$ sh ~/rust-installer/rustup.sh --prefix=~/rust --spec=beta -y --disable-sudo 2> /dev/null

rustup: gpg not available. signatures will not be verified

rustup: downloading manifest for 'beta'

rustup: downloading toolchain for 'beta'

No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.

Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
```

abonander commented:
Sticking with `try!()` is for compatibility with older stable Rust releases. I'm not sure how many people are still stuck on old stable releases and I'm not sure it's possible to tell without surveying the community, but it's just something I'm concerned about. I'm also sure that I've broken compatibility somewhere else anyways by using newer APIs, but that's beside the point. The `?` operator bothers me a little bit at a fundamental level, it's too easy to miss given that it participates in control flow.

little-dude commented:
Alright, I switched to `try!`.

abonander commented:
So I tested it real quick, and besides some issues with macro hygiene and import shadowing in `src\local_test.rs`, the current master branch builds and tests with Rust 1.12.0 (`?` was stabilized in 1.13.0). I've got to see how far back it still works, I'm pleasantly surprised.

abonander commented:
Eh, it doesn't compile with 1.11 because `impl<T> From<T> for Option<T>` wasn't done yet, and that's a major part of the API that I'd rather not change, so unless there's a large number of users stuck on 1.12, I don't see much harm in adopting the `?` operator across the board.

little-dude commented:
> the current master branch builds and tests with Rust 1.12.0

Cool :)

> Addendum: that's not necessary here but if you want to open a follow-up PR doing that, that'd be cool too.

Will do, then!

abonander commented on src/client/lazy.rs:
```diff
@@ -507,6 +507,7 @@ mod hyper {
     impl<'d> super::PreparedFields<'d> {
         /// #### Feature: `hyper`
         /// Convert `self` to `hyper::client::Body`.
+        #[cfg_attr(feature="clippy", allow(wrong_self_convention))]
```
Let it warn on this for now, that way I don't forget to fix this for the next minor version release.

little-dude replied:
done in 019b645

abonander commented on src/server/field.rs:
```diff
@@ -556,10 +522,7 @@ pub trait ReadEntry: PrivReadEntry + Sized {
-            Some(headers) => headers,
-            None => return End(self),
-        };
+        let field_headers = try_read_entry!(self; self.read_headers().map_err(|e| io::Error::new(io::ErrorKind::Other, e)));
```
I'm not sure of the convention here, but I think `io::ErrorKind::InvalidData` is more appropriate.

abonander replied:
This could also just be done in an `Into<io::Error>` impl I think.

abonander commented on src/server/field.rs:
```diff
@@ -62,13 +51,11 @@ where R: BufRead, F: FnOnce(&[StrHeader]) -> Ret {
-            error!("Could not read field headers.");
-            // RFC: return an actual error instead?
-            return Ok(closure(&[]));
+            return Err(ParseHeaderError::Other("Unknown error while parsing headers".to_string()));
```
So this would only be hit if `fill_buf()` returned too small a buffer too many times, it's kind of relying on `BoundaryReader::read_to_boundary()` getting more data each attempt but that's outside of convention for `BufRead` (all standard impls either return all their data at once or don't read more until the buffer is completely consumed). This loop should only need to repeat once, if at all. So it's not really an error, the stream is just behaving weirdly and we don't want to loop forever because that could be used to DoS the server running this.

abonander commented on src/server/field.rs:
```diff
@@ -138,63 +125,42 @@ pub struct ContentDisp {
-                return None;
-            };
-
+                return Err(ParseHeaderError::Invalid);
```
I'd like to have this contain the error message that was previously being logged, otherwise there's not much context for the error.

little-dude replied:
see 8ff0962
I transformed the logs into a message carried by the error.

abonander commented on src/server/field.rs:
```diff
@@ -138,63 +125,42 @@ pub struct ContentDisp {
-        find_header(headers, CONTENT_TYPE),
-        debug!("Content-Type header not found for field.")
-    );
+    let header = try!(find_header(headers, CONTENT_TYPE).ok_or(ParseHeaderError::Missing));
```
A `Content-Type` header isn't required, it just implies a file field.

abonander commented on src/server/field.rs:
```diff
@@ -138,63 +125,42 @@ pub struct ContentDisp {
+    fn parse(headers: &[StrHeader]) -> Result<ContentDisp, ParseHeaderError> {
         if headers.is_empty() {
-            return None;
+            return Err(ParseHeaderError::Missing);
```
This isn't an error, there's just no more fields. A missing error would be if `headers` wasn't empty, but `Content-Disposition` (required header) wasn't found.

little-dude replied:
I think 781646c addresses this.

abonander commented on Cargo.toml:
```diff
@@ -21,6 +21,7 @@ mime = "0.2"
 mime_guess = "1.8"
 rand = "0.3"
 tempdir = ">=0.3.4"
+clippy = { version = ">=0.0, <0.1", optional = true}
```
You could add `clippy` to the `nightly` feature so it gets tested, if you like.

abonander commented:
So 1.13 needed the changes to `src/local_test.rs` too, which I've got on the `rust-1.13` if you want to fork from that for the `?` operator conversion. Since it didn't previously compile that far back, pinning the version isn't a breaking change, but maybe it should still be done in a minor version bump. What do you think?

abonander commented on src/server/field.rs:
```diff
@@ -714,4 +677,56 @@ impl<M: ReadEntry, Entry> ReadEntryResult<M, Entry> {
+#[derive(Debug)]
+pub enum ParseHeaderError {
+    /// The header was not found
+    Missing,
```
I'd like this variant to say which header was missing, though only `Content-Disposition` is required so maybe that can be hardcoded.

Also, this enum doesn't necessarily have to be public if it's converted to `io::Error` first, which ends up happening anyway.

little-dude replied:
see 	f429ee7 that introduces an enum to tell which header could not be found,
and 2d95162 that makes this enum and the whole `ParseHeaderError` private.

abonander replied:
👍

abonander commented on src/lib.rs:
```diff
@@ -30,7 +30,7 @@
 #![cfg_attr(feature="clippy", feature(plugin))]
 #![cfg_attr(feature="clippy", plugin(clippy))]
-#![cfg_attr(feature="clippy", deny(clippy))]
+#![cfg_attr(feature="clippy", warn(clippy))]
```
Does it not work to leave `deny` here but `warn` on the one function? It's not a big deal anyway.

little-dude replied:
ah yes, let me amend this one.

abonander commented on src/server/field.rs:
```diff
@@ -714,4 +688,63 @@ impl<M: ReadEntry, Entry> ReadEntryResult<M, Entry> {
+#[derive(Debug)]
+enum ParseHeaderError {
+    /// The header was not found
+    NotFound(HeaderType),
```
Yeah `MissingContentDisposition` here would be good.

little-dude replied:
done in 3c00a62

little-dude commented:
I think I addressed your comments. I can squash some commits if you want.
I'll make a separate PR for the `?` conversion.

>  pinning the version isn't a breaking change, but maybe it should still be done in a minor version bump.

I agree 👍

abonander commented:
I'm not too worried about squashing, the commits are pretty self-contained as-is.

abonander commented:
Thanks for the hard work! It's much appreciated.

abonander commented:
@little-dude I think I'm going to move Nickel integration to its own crate (in the same repo). I'm going to have that change come before the minor version bump (and thus the `?` conversion) so users of `0.10` see the in-crate integration deprecated for at least one release first.

little-dude commented:
Alright :) And thanks for guiding me throughout this PR.
